File: src/nougen_shards/gatekeeper.py
"""
DavOs Gatekeeper: Security middleware enforcing Mutation Gates.
Determines if actions or commands violate the constitution.
"""
import re

def check_mutation_gate(command: str, parameters: dict = None) -> dict:
    """
    Checks if a proposed command or parameter dict hits a locked mutation gate.
    
    Args:
        command: The command string (e.g., shell command, database query).
        parameters: Optional dictionary of parameters (e.g., {"dry_run": False}).
        
    Returns:
        dict: {"allowed": bool, "reason": str, "gate": str}

    NOTE: This is a defense-in-depth speed-bump, NOT a security boundary.
    It is a best-effort denylist meant to catch obvious destructive commands
    and slow down accidents; it can be trivially bypassed by obfuscation
    (encoding, indirection, aliases, etc.) and must never be relied upon as
    the sole protection against malicious input.
    """
    # Normalize before matching so trivial whitespace/case obfuscation
    # (e.g. "rm   -rf", "git\tpush", "RM -RF") is still caught.
    cmd_lower = re.sub(r"\s+", " ", command.lower()).strip()

    # 1. Database Schema or Index Alteration Gate
    schema_patterns = [
        r"\balter\s+table\b",
        r"\bcreate\s+table\b",
        r"\bdrop\s+table\b",
        r"\bcreate\s+index\b",
        r"\bdrop\s+index\b",
    ]
    for pattern in schema_patterns:
        if re.search(pattern, cmd_lower):
            return {
                "allowed": False,
                "reason": "Database schema or index modifications are restricted to GM approval.",
                "gate": "schema_change"
            }

    # 2. Destructive Cleanup Gate
    destructive_patterns = [
        # Filesystem destruction (shell)
        r"\brm\s+-[rf]+\b",                 # rm -rf / rm -fr / rm -r / rm -f
        r"\brm\s+--recursive\b",
        r"\brm\s+--force\b",
        r"\bdel\s+/",                       # Windows: del /s /q
        r"\bformat\s",                      # Windows: format c:
        r"\bmkfs\b",                        # make filesystem (wipes device)
        r"\bdd\s+if=",                      # raw disk overwrite
        r"\btruncate\b",                    # truncate file / SQL TRUNCATE
        r">\s*/dev/sd[a-z]",                # overwrite raw block device
        r":\(\)\s*\{",                      # forkbomb :(){ :|:& };:
        # Filesystem destruction (Python)
        r"\bshutil\.rmtree\b",
        r"\bos\.remove\b",
        r"\bos\.unlink\b",
        # Destructive SQL
        r"\bdelete\s+from\b",
        r"\bdrop\s+database\b",
    ]
    for pattern in destructive_patterns:
        if re.search(pattern, cmd_lower):
            return {
                "allowed": False,
                "reason": "Destructive cleanups and table deletions are restricted to GM approval.",
                "gate": "destructive_cleanup"
            }

    # 3. Dry-run False Gate
    if parameters and parameters.get("dry_run") is False:
        return {
            "allowed": False,
            "reason": "Actions with dry_run=False require GM approval.",
            "gate": "dry_run_false"
        }

    # 4. Paid / Billing / Quota Gate
    billing_patterns = [
        r"\bbilling\b",
        r"\bquota\b",
        r"\bpaid-tier\b",
        r"\bsubscription\b",
        r"\bbudget\b",
    ]
    for pattern in billing_patterns:
        if re.search(pattern, cmd_lower):
            return {
                "allowed": False,
                "reason": "Billing, quota, and paid-tier modifications require GM approval.",
                "gate": "billing_quota_paid_tier_change"
            }

    # 5. Deployment & Registry Gate
    deploy_patterns = [
        r"\bgit\s+push\b",                  # covers `git push --force` / `git push -f`
        r"\bgit\s+reset\s+--hard\b",
        r"\bnpm\s+publish\b",
        r"\bdeploy\b",
        r"\bregister-node\b",
        r"\bchmod\s+-r\s+777\b",            # recursive world-writable perms
    ]
    for pattern in deploy_patterns:
        if re.search(pattern, cmd_lower):
            return {
                "allowed": False,
                "reason": "Deployment actions and node registration changes require GM approval.",
                "gate": "deployment_target_change"
            }

    return {
        "allowed": True,
        "reason": "Action is non-mutating and safe to execute.",
        "gate": None
    }

File: ts/src/nougen_shards/gatekeeper.ts
/**
 * DavOs Gatekeeper: Security middleware enforcing Mutation Gates. (TS mimic of gatekeeper.py)
 * Determines if actions or commands violate the constitution.
 */

export interface GateResult {
  allowed: boolean;
  reason: string;
  gate: string | null;
}

/**
 * Checks if a proposed command or parameter dict hits a locked mutation gate.
 *
 * @param command The command string (e.g., shell command, database query).
 * @param parameters Optional dictionary of parameters (e.g., {"dry_run": false}).
 * @returns {allowed, reason, gate}
 *
 * NOTE: This is a defense-in-depth speed-bump, NOT a security boundary.
 * It is a best-effort denylist meant to catch obvious destructive commands
 * and slow down accidents; it can be trivially bypassed by obfuscation
 * (encoding, indirection, aliases, etc.) and must never be relied upon as
 * the sole protection against malicious input.
 */
export function check_mutation_gate(
  command: string,
  parameters: Record<string, any> | null = null,
): GateResult {
  // Normalize before matching so trivial whitespace/case obfuscation
  // (e.g. "rm   -rf", "git\tpush", "RM -RF") is still caught.
  const cmd_lower = command.toLowerCase().replace(/\s+/g, " ").trim();

  // 1. Database Schema or Index Alteration Gate
  const schema_patterns = [
    /\balter\s+table\b/,
    /\bcreate\s+table\b/,
    /\bdrop\s+table\b/,
    /\bcreate\s+index\b/,
    /\bdrop\s+index\b/,
  ];
  for (const pattern of schema_patterns) {
    if (pattern.test(cmd_lower)) {
      return {
        allowed: false,
        reason: "Database schema or index modifications are restricted to GM approval.",
        gate: "schema_change",
      };
    }
  }

  // 2. Destructive Cleanup Gate
  const destructive_patterns = [
    // Filesystem destruction (shell)
    /\brm\s+-[rf]+\b/, // rm -rf / rm -fr / rm -r / rm -f
    /\brm\s+--recursive\b/,
    /\brm\s+--force\b/,
    /\bdel\s+\//, // Windows: del /s /q
    /\bformat\s/, // Windows: format c:
    /\bmkfs\b/, // make filesystem (wipes device)
    /\bdd\s+if=/, // raw disk overwrite
    /\btruncate\b/, // truncate file / SQL TRUNCATE
    />\s*\/dev\/sd[a-z]/, // overwrite raw block device
    /:\(\)\s*\{/, // forkbomb :(){ :|:& };:
    // Filesystem destruction (Python)
    /\bshutil\.rmtree\b/,
    /\bos\.remove\b/,
    /\bos\.unlink\b/,
    // Destructive SQL
    /\bdelete\s+from\b/,
    /\bdrop\s+database\b/,
  ];
  for (const pattern of destructive_patterns) {
    if (pattern.test(cmd_lower)) {
      return {
        allowed: false,
        reason: "Destructive cleanups and table deletions are restricted to GM approval.",
        gate: "destructive_cleanup",
      };
    }
  }

  // 3. Dry-run False Gate
  if (parameters && parameters["dry_run"] === false) {
    return {
      allowed: false,
      reason: "Actions with dry_run=False require GM approval.",
      gate: "dry_run_false",
    };
  }

  // 4. Paid / Billing / Quota Gate
  const billing_patterns = [
    /\bbilling\b/,
    /\bquota\b/,
    /\bpaid-tier\b/,
    /\bsubscription\b/,
    /\bbudget\b/,
  ];
  for (const pattern of billing_patterns) {
    if (pattern.test(cmd_lower)) {
      return {
        allowed: false,
        reason: "Billing, quota, and paid-tier modifications require GM approval.",
        gate: "billing_quota_paid_tier_change",
      };
    }
  }

  // 5. Deployment & Registry Gate
  const deploy_patterns = [
    /\bgit\s+push\b/, // covers `git push --force` / `git push -f`
    /\bgit\s+reset\s+--hard\b/,
    /\bnpm\s+publish\b/,
    /\bdeploy\b/,
    /\bregister-node\b/,
    /\bchmod\s+-r\s+777\b/, // recursive world-writable perms
  ];
  for (const pattern of deploy_patterns) {
    if (pattern.test(cmd_lower)) {
      return {
        allowed: false,
        reason: "Deployment actions and node registration changes require GM approval.",
        gate: "deployment_target_change",
      };
    }
  }

  return {
    allowed: true,
    reason: "Action is non-mutating and safe to execute.",
    gate: null,
  };
}
